### ISSUE TITLE
Developer Cadence in Updating a Module

### ISSUE THREAD

**Issue by codeputer** [Azure/iotedge#7041]
Shorten the dev cycle to build and deploy a new image from Visual Studio (NOT VS Code).
My current steps (which work but is tedious to repeat)
1) Ensure the module code is rebuilt
2) Build and Push the IOT Edge Module (I check the SHA to ensure that my visual studio version is registered).
3) Ensure the module (not device) is deleted from Azure via the IOT Hub device that is sitting next to me on the desktop. Wait for Agent to remove the previous deployment.
4) Jump to Terminal, and issue this command - "az iot edge set-module...." (which is NOT available in VS IOT Edge plugin, and I learned late in my development)
5) return to Azure and the module appears
6) wait for the module properties to be updated (manual refresh)
7) navigate to troubleshooting
8) download feedback and review
9) make changes to the program
10) repeat step 1

my preference would be:
Step 1, 2 and then on Step 3, on the Device, "redeploy custom modules".... and SHOW THE SHA of the image which was retrieved from the Azure Registry.

I tried the IOTEdgeFlow locally (big rabbit hole due to too many SSH device authentications on my desktop), and had my program working using a fake implementation of my dependencies to the board.  But that didn't find the Ansi(0) character appended to commands like "cat /", or permission issues to find GPIO and storage paths that are required.  There are times you just need to run on the device.

I'm developing a custom module and the learning curve around IOT Edge, IOT Agent, IOT Hub, Devices, IOT Device, Device Twins, Deployment Manifests, Docker, and Docker build scripts was a significate rabbit hole that I was not prepared for!

For reference, I had this program running on the device (raspberry pi) as a native application (rather than a docker image).  Then I realized that I needed to deploy over the air to devices that are NOW being shipped around the world rather than locally.  Considered a custom integration with two programs, but figured I should use the existing services of IOT Hub, Device Edge Hub and Agent - but I'm not sure now.

**Comment by david-emakenemi** [comment]
Hey, @codeputer thanks for reaching out! I'm slightly confused, what is the high-level goal you are trying to achieve?

**Comment by codeputer** [comment]
Currently, I listed a 10-step process in which to update a physical device, as some things just can't be tested on a simulated device.  I just want to limit the number of steps.  Why should I have to delete the module and then do a push, why not a redo deployment button on the set module screen?

**Comment by david-emakenemi** [comment]
hmm, by "update a physical device", are you trying to test new versions of the custom module you developed? If so, you can use stable tags: https://learn.microsoft.com/en-us/azure/container-registry/container-registry-image-tag-version#stable-tags
But for this, you'll have to get the versioning right.

The steps are:
- Build an image on your box
- Push the image to azure
- Have your deployment auto-detect that there's a new image where you can auto-deploy to wherever

**Comment by david-emakenemi** [comment]
Suppose that doesn't answer your question, can you clarify what problem you are trying to solve? Are you trying to develop custom modules and use IoT Edge for orchestration?

**Comment by codeputer** [comment]
in short - yes that is what I'm attempting to do..."develop custom modules and use IoT Edge for orchestration"

It's perhaps a learning curve issue... a bit of context. I created a program that runs well on the native device, and sends messages to IOT Hub.  My world was good until I started to think about using IOT Hub edge Management features of the program, which meant I needed to make it a "module", and I didn't realize just how deep this rabbit hole was.

Using a "module", I'm having issues interacting with the Raspberry PI 2  Zero board, in terms of HostNaming the computer, GPIO, and storage access.  Most of which I have resolved over a very long and painful learning process.   Each time I "Build IOT Image and Push" to the registry, in my case, does NOT give the IOT Agent an indication that a new image needs to be deployed.  This I think goes to your comment on Tagging.

The documentation, as I have reviewed, allows for a one-time deployment, as well as the use of EFLOW to "simulate the device", but it does not discuss adequately the cadence of build, registry update, tag update, and finally the deployment, all undera a timeline that needs to happen repeatedly (especially during the steep learning curve).

I have yet to understand the association between the registry update, and the deployment of a version of the application which is in development.  This is a different use case than the deployment of a stable application, under a new version number, and deployment across a wide set of devices.  In fact, I still don't know how to "replace" a version, other than to "add" a new module, and "remove" the old module(?).

The Tagging documentation that I have reviewed, is a manual process of version update in the module.json, as well as the deployment.template.json - which I think needs to be aligned to work correctly.  if this versioning is updated correctly (and yet another deployment to Azure), then I think the IOT Edge Agent, as it contacts the host, would notice the version change and initiate a new deployment.  I compare this to simply going to "Set Modules", deleting it, waiting for removal, and then adding it back (again, outside of Visual Studio, and executing the "az iot edge set-module" command (which minimizes the typing).  That said, why this is not an option in visual studio is not something I cannot understand - to execute this, a subscription must be selected (which I have many), an IOT hub name and connection, a configuration file, a device id - all of which is known by Visual Studio.  (I understand VS Code is better, but that is not my IDE of choice, as my focus is on the C# and not the widespread plugin capabilities of VS Code across so many different languages.)

In short (sorry for the long format - vent?), I need an image that will host my program - I selected 7.0-bullseye-slim-arm64v8 and gave it (for now) privileged access to the host (GPIO and storage).  Lots of learning here, but the latest is that /usr/bin of the image is not the same as /usr/bin of the host.  (I need raspi-config nonint capabilities).  I'm learning how the Docker Build script creates the docker image, the capabilities within the build script to prepare the "image" correctly - but after a month duration (time-sliced as I have several clients), I'm wondering if this is all worth it.

(unrelated but important) I will have thousands of devices in the next six months, deployed worldwide, and I need to manage them.  In the assembly of the units, running a command like (sudo iotedge config mp --connection-string 'PASTE_DEVICE_CONNECTION_STRING_HERE') on the device, after it boots up, is not something I think can be sustained by the resources that are assembling the units.  My preference is to deploy an image that will use the "serial number" of the device to query the Registry Manager to self-discover the "device id" associated with that serial number.  It will then needs to "become" that device (HostName change, network name change, and also an "iotedge config mp" change (I pass the device connection string in the Query of the Registry Manager).  I don't understand the configuration of each device, in the style that IOT Hub documents, as possible on even this scale of deployment.
